### PR TITLE
Replace urllib by requests in contentproviders

### PR DESCRIPTION
**Pull request by douardda**
requests is globally simpler to use, and more and more people
are more familiar with this later than urllib.

See #988 for a discussion on why this is needed.

### CODE REVIEW

**Comment by douardda** [comment]
Any chance to have this PR being merged? Thanks

**Comment by douardda** [comment]
@betatim @manics ping?

**Review by betatim — COMMENTED** [review]
Sorry for the super slow response. I ended up taking some time off from maintaining duties to recharge.

I left a few nit picks but otherwise this looks like a huge effort and as far as I can tell it doesn't change anything in the tests besides switching the mock. This is a good thing :)

File: repo2docker/contentproviders/doi.py [hunk 1]
```diff
@@ -53,38 +68,42 @@ def fetch_file(self, file_ref, host, output_dir, unzip=False):
...
+
+        if path.dirname(fname):
+            sub_dir = path.join(output_dir, path.dirname(fname))
+            if not path.exists(sub_dir):
+                yield "Creating {}\n".format(sub_dir)
+                makedirs(sub_dir, exist_ok=True)
+
+        dst_fname = path.join(output_dir, fname)
+        with open(dst_fname, "wb") as dst:
+            yield "Fetching {}\n".format(fname)
```
[the remark this answers is not in the document]
Comment (douardda) [reply repo2docker/contentproviders/doi.py:84]: Good question. Previsouly, this fetchfile() method did log (as generated strings):
- the directory creation ("Creating [...]"),
- the saving of the downloaded zip file in `dst_fname` ("Fetching [...]")
- the zip file extraction ("Extracting [...]")
- the list of extracted files ("Extracted [...]")

In my version of the code, the only thing logged in addition to these is the "Requestion [...]" line in which I display the actually retrieved URL (after doi resolution) I though would be useful in a log.

File: repo2docker/contentproviders/doi.py [hunk 2]
```diff
@@ -53,38 +68,42 @@ def fetch_file(self, file_ref, host, output_dir, unzip=False):
...
+
+            new_subdirs = os.listdir(output_dir)
+            # if there is only one new subdirectory move its contents
+            # to the top level directory
+            if len(new_subdirs) == 1:
+                d = new_subdirs[0]
+                copytree(path.join(output_dir, d), output_dir)
+                shutil.rmtree(path.join(output_dir, d))
+
+            yield "Fetched files: {}\n".format(os.listdir(output_dir))
```
[the remark this answers is not in the document]
Comment (douardda) [reply repo2docker/contentproviders/doi.py:109]: This `yield "Fectched Files:"` is not new; it was already there before.

File: repo2docker/contentproviders/hydroshare.py [hunk 3]
```diff
@@ -16,9 +16,8 @@ class Hydroshare(DoiProvider):
 
     def _fetch_version(self, host):
         """Fetch resource modified date and convert to epoch"""
-        json_response = json.loads(
-            self.urlopen(host["version"].format(self.resource_id)).read()
-        )
+        json_response = self.urlopen(host["version"].format(self.resource_id)).json()
+        print("json_response=", json_response)
```
[the remark this answers is not in the document]
Comment (douardda) [reply repo2docker/contentproviders/hydroshare.py]: updated the PR to remove this print statement

**Pull request merged**